crates/turborepo-filewatch/src/package_watcher.rs

```diff
@@ -0,0 +1,215 @@
+    package_data: Arc<Mutex<HashMap<AbsoluteSystemPathBuf, PackageData>>>,
+}
+
+impl PackageWatcher {
+    /// Creates a new package watcher whose current package data can be queried.
+    pub fn new<T: PackageDiscovery + Send + 'static>(
+        root: AbsoluteSystemPathBuf,
+        recv: broadcast::Receiver<Result<Event, NotifyError>>,
+        discovery: T,
+    ) -> Result<Self, package_manager::Error> {
+        let (exit_tx, exit_rx) = oneshot::channel();
+        let package_data: Arc<Mutex<_>> = Default::default();
+        let subscriber = Subscriber::new(exit_rx, root, package_data.clone(), recv, discovery)?;
+        let handle = tokio::spawn(subscriber.watch());
+        Ok(Self {
+            _exit_tx: exit_tx,
+            _handle: handle,
+            package_data,
+        })
+    }
+
+    pub async fn get_package_data(&self) -> Vec<PackageData> {
+        self.package_data
+            .lock()
+            .await
+            .clone()
+            .into_values()
+            .collect()
+    }
+}
+
+/// The underlying task that listens to file system events and updates the
+/// internal package state.
+struct Subscriber<T: PackageDiscovery> {
+    exit_rx: oneshot::Receiver<()>,
+    filter: WorkspaceGlobs,
+    recv: broadcast::Receiver<Result<Event, NotifyError>>,
+    package_data: Arc<Mutex<HashMap<AbsoluteSystemPathBuf, PackageData>>>,
+    manager: PackageManager,
+    repo_root: AbsoluteSystemPathBuf,
+    discovery: T,
+}
+
+impl<T: PackageDiscovery + Send + 'static> Subscriber<T> {
+    fn new(
+        exit_rx: oneshot::Receiver<()>,
+        repo_root: AbsoluteSystemPathBuf,
+        package_data: Arc<Mutex<HashMap<AbsoluteSystemPathBuf, PackageData>>>,
+        recv: broadcast::Receiver<Result<Event, NotifyError>>,
+        discovery: T,
+    ) -> Result<Self, Error> {
+        let manager = PackageManager::get_package_manager(&repo_root, None)?;
+
+        Ok(Self {
+            exit_rx,
+            filter: manager.get_workspace_globs(&repo_root)?,
+            package_data,
+            recv,
+            manager,
+            repo_root,
+            discovery,
+        })
+    }
+
+    async fn watch(mut self) {
+        // initialize the contents
+        self.rediscover_packages().await;
+
+        // respond to changes
+        loop {
+            tokio::select! {
+                _ = &mut self.exit_rx => return,
+                file_event = self.recv.recv().into_future() => match file_event {
+                    Ok(Ok(event)) => self.handle_file_event(event).await,
+                    // if we get an error, we need to re-discover the packages
+                    Ok(Err(_)) => self.rediscover_packages().await,
+                    Err(RecvError::Closed) => return,
+                    // if we end up lagging, just skip the event
+                    Err(RecvError::Lagged(count)) => {
+                        tracing::warn!("lagged behind {count} processing file watching events");
```

Comment:
I think we should consider this the same as receiving a filewatching error and re-discover packages.